### PR TITLE
add attachments to evaluate

### PR DESCRIPTION
**Pull request by isahers1**

### CODE REVIEW

File: python/tests/integration_tests/test_client.py [hunk 1]
```diff
@@ -1160,20 +1157,18 @@ def test_list_examples_attachments_keys(langchain_client: Client) -> None:
...
 
-@pytest.mark.skip(
-    reason="Need to land https://github.com/langchain-ai/langsmith-sdk/pull/1209 first"
-)
 def test_evaluate_with_attachments(langchain_client: Client) -> None:
     """Test evaluating examples with attachments."""
     dataset_name = "__test_evaluate_attachments" + uuid4().hex[:4]
-    # 1. Create dataset
+    langchain_client = Client(
+        api_key="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",
```
Comment (hinthornw) [inline python/tests/integration_tests/test_client.py:1164]: You ought to delete this key now

**Comment by isahers1** [comment]
need to add to summary evaluators as well I think

File: python/tests/integration_tests/test_client.py [hunk 2]
```diff
@@ -1381,6 +1370,106 @@ def evaluator(run: Run, example: Example) -> Dict[str, Any]:
...
+        return {"answer": "4"}
+
+    async def evaluator(run: Run, example: Example) -> Dict[str, Any]:
+        return {
+            "score": float(
+                run.outputs.get("answer") == example.outputs.get("answer")  # type: ignore
+            )
+        }
+
+    results = await aevaluate(
```
Comment (baskaryan) [inline python/tests/integration_tests/test_client.py:1464]: can now do langchain_client.aevaluate

File: python/langsmith/client.py [hunk 3]
```diff
@@ -140,6 +140,16 @@ class ZoneInfo:  # type: ignore[no-redef]
 URLLIB3_SUPPORTS_BLOCKSIZE = "key_blocksize" in signature(PoolKey).parameters
 
 
+class AutoSeekBytesIO(io.BytesIO):
```
Comment (baskaryan) [inline python/langsmith/client.py:143]: is this related?

**Review by baskaryan — COMMENTED** [review]
do we need to inspect the evaluator signature ahead of time so that if user passes in just a dataset name/id, we know if we need to load attachments?

File: python/langsmith/client.py (hunk 3 above)
Comment (isahers1) [reply python/langsmith/client.py:143]: no I need to remove this, this was to solve the issue that reading from the attachment in the target function made it unaccessible (until you seek back to 0) in the evaluator, but we solved this in a different way

File: python/langsmith/evaluation/evaluator.py [hunk 4]
```diff
@@ -624,7 +624,14 @@ def _normalize_evaluator_func(
...
     Callable[[Run, Optional[Example]], Awaitable[_RUNNABLE_OUTPUT]],
 ]:
-    supported_args = ("run", "example", "inputs", "outputs", "reference_outputs")
+    supported_args = (
+        "run",
+        "example",
+        "inputs",
+        "outputs",
+        "reference_outputs",
+        "attachments",
```
Comment (baskaryan) [inline python/langsmith/evaluation/evaluator.py:633]: should we rename input_attachments

**Review by agola11 — APPROVED** [review]

File: python/langsmith/evaluation/_runner.py [hunk 5]
```diff
@@ -1913,7 +1914,30 @@ def _ensure_traceable(
...
+        return False
+    return any(
+        any(
+            p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD)
+            and p.name == "attachments"
+            for p in (
+                inspect.signature(
+                    e.__call__ if hasattr(e, "__call__") else e
+                ).parameters.values()
+                if callable(e) or hasattr(e, "__call__")
```
Comment (baskaryan) [inline python/langsmith/evaluation/_runner.py:1930]: ooc what kinds of objects have `__call__` but aren't callable?

**Review by baskaryan — APPROVED** [review]
